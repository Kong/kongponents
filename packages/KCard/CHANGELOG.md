# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 6.9.0 (2022-02-04)


### Features

* **kpop:** add positionFixed prop ([#481](https://github.com/Kong/kongponents/issues/481)) ([62dcbe6](https://github.com/Kong/kongponents/commit/62dcbe6cf6c82bf628316e3328f784cf32308eec))





## 6.8.1 (2022-02-03)

**Note:** Version bump only for package @kongponents/kcard





# 6.8.0 (2022-02-03)


### Features

* **ktable:** add legacy ktable component ([#477](https://github.com/Kong/kongponents/issues/477)) ([4454617](https://github.com/Kong/kongponents/commit/445461705d25b27426e2b542e65e8ad9a351a921))





## 6.7.11 (2022-01-21)

**Note:** Version bump only for package @kongponents/kcard





## 6.7.10 (2022-01-12)


### Bug Fixes

* **ktable:** update sort styling ([#473](https://github.com/Kong/kongponents/issues/473)) ([d2d242f](https://github.com/Kong/kongponents/commit/d2d242fbeaf92db112f699cbe8eed163b066feff))





## 6.7.9 (2022-01-07)

**Note:** Version bump only for package @kongponents/kcard





## 6.7.8 (2022-01-07)


### Bug Fixes

* **kinput:** allow numbers to be passed in as value ([#470](https://github.com/Kong/kongponents/issues/470)) ([fbf4c00](https://github.com/Kong/kongponents/commit/fbf4c00291383707770076110dcfa9519d31faef))





## 6.7.7 (2021-12-21)


### Reverts

* **version:** revert independent version fix ([#469](https://github.com/Kong/kongponents/issues/469)) ([b647e56](https://github.com/Kong/kongponents/commit/b647e56b5c05c25e55362a845e1c4a30fe40d910))





## 6.7.6 (2021-12-21)

**Note:** Version bump only for package @kongponents/kcard





## 6.7.3 (2021-12-20)

**Note:** Version bump only for package @kongponents/kcard





## 6.7.2 (2021-12-17)


### Bug Fixes

* **actions:** fix actions ([#459](https://github.com/Kong/kongponents/issues/459)) ([9bb5ad5](https://github.com/Kong/kongponents/commit/9bb5ad57272f9f57c140646eb9d4413d595115d6))





## 6.7.1 (2021-12-17)

**Note:** Version bump only for package @kongponents/kcard





# 6.7.0 (2021-12-17)


### Features

* **kinput:** add css variable for optional kinput placeholder text color overrides ([#443](https://github.com/Kong/kongponents/issues/443)) ([3b40207](https://github.com/Kong/kongponents/commit/3b40207eabd87b21568ec2697ad39961c2a91b40))





## 6.6.5 (2021-12-17)


### Bug Fixes

* **kpagination:** fix currentPage prop ([#455](https://github.com/Kong/kongponents/issues/455)) ([5e1a801](https://github.com/Kong/kongponents/commit/5e1a801114d2357a2d49dca3bb2571dd055b4119))





## 6.6.4 (2021-12-16)


### Bug Fixes

* **ktextarea:** fix v-model ([#454](https://github.com/Kong/kongponents/issues/454)) ([35f581f](https://github.com/Kong/kongponents/commit/35f581f5c1be13fdebf1a63b8742181de8e19a8b))





## 6.6.3 (2021-12-15)

**Note:** Version bump only for package @kongponents/kcard





## 6.6.2 (2021-12-10)

**Note:** Version bump only for package @kongponents/kcard





## 6.6.1 (2021-12-10)


### Bug Fixes

* **kalert:** fix invalid html structure ([#452](https://github.com/Kong/kongponents/issues/452)) ([15eade8](https://github.com/Kong/kongponents/commit/15eade87d0661ae7716a6e32f1fa63279dbf8743))





# 6.6.0 (2021-12-08)


### Features

* **comment:** add global comment in output file ([#451](https://github.com/Kong/kongponents/issues/451)) ([b2f6154](https://github.com/Kong/kongponents/commit/b2f6154c37be30a3d364337f3d87fe40a64fc753))





## 6.5.9 (2021-12-01)


### Bug Fixes

* infinite publish ([d3f275b](https://github.com/Kong/kongponents/commit/d3f275bb668dd95782f6150232cc22a5b95b1bcf))





## [6.5.8](https://github.com/Kong/kongponents/compare/@kongponents/kcard@6.5.7...@kongponents/kcard@6.5.8) (2021-12-01)

**Note:** Version bump only for package @kongponents/kcard





## [6.5.7](https://github.com/Kong/kongponents/compare/@kongponents/kcard@6.5.6...@kongponents/kcard@6.5.7) (2021-12-01)

**Note:** Version bump only for package @kongponents/kcard





## [6.5.6](https://github.com/Kong/kongponents/compare/@kongponents/kcard@6.5.5...@kongponents/kcard@6.5.6) (2021-12-01)

**Note:** Version bump only for package @kongponents/kcard





## [6.5.5](https://github.com/Kong/kongponents/compare/@kongponents/kcard@6.5.4...@kongponents/kcard@6.5.5) (2021-12-01)

**Note:** Version bump only for package @kongponents/kcard





## 6.5.4 (2021-12-01)

**Note:** Version bump only for package @kongponents/kcard





## 6.5.3 (2021-11-30)

**Note:** Version bump only for package @kongponents/kcard





## 6.5.2 (2021-11-29)


### Bug Fixes

* **ktable:** allow init client sort ([c3e2a69](https://github.com/Kong/kongponents/commit/c3e2a693974d6aae7f829841af89391c5dd357b2))





## 6.5.1 (2021-11-24)


### Bug Fixes

* **ktable/kcardcatalog:** fetch after init ([#442](https://github.com/Kong/kongponents/issues/442)) ([4a053ae](https://github.com/Kong/kongponents/commit/4a053ae4e39e7df7acc798510c8a2fa9e5ab3f6c))





# 6.5.0 (2021-11-24)


### Features

* **ktable:** add cacheKey prop ([#441](https://github.com/Kong/kongponents/issues/441)) ([743082d](https://github.com/Kong/kongponents/commit/743082df383699b8016d3b6d85497ec77563aaec))





# 6.4.0 (2021-11-24)


### Features

* **ktable:** default sort to asc ([#440](https://github.com/Kong/kongponents/issues/440)) ([95f2e91](https://github.com/Kong/kongponents/commit/95f2e9164c09e50b25f2e334ff30231ab0984503))





# 6.3.0 (2021-11-23)


### Features

* **ktable/kcardcatalog:** make states slottable ([#438](https://github.com/Kong/kongponents/issues/438)) ([43eac93](https://github.com/Kong/kongponents/commit/43eac9337193fe47b28b27576d727e6cbbabad50))





# 6.2.0 (2021-11-23)


### Features

* **ktable/kcardcatalog:** emit cta click events ([#437](https://github.com/Kong/kongponents/issues/437)) ([29eb2a5](https://github.com/Kong/kongponents/commit/29eb2a53cf180935afee1ee10f1b88b462dc29ad))





## 6.1.19 (2021-11-23)

**Note:** Version bump only for package @kongponents/kcard





## 6.1.18 (2021-11-23)


### Bug Fixes

* **ktable:** add missing style ([#435](https://github.com/Kong/kongponents/issues/435)) ([8782dee](https://github.com/Kong/kongponents/commit/8782dee82dafc68c1f956149a249b950359401b1))





## 6.1.17 (2021-11-22)


### Bug Fixes

* fix deps and build process ([#434](https://github.com/Kong/kongponents/issues/434)) ([9fdd180](https://github.com/Kong/kongponents/commit/9fdd180c9816bd4a43d4cd9229fc45b180ba5eed))





## 6.1.16 (2021-11-19)


### Bug Fixes

* **publish:** update node version in publish ([#433](https://github.com/Kong/kongponents/issues/433)) ([26cfe34](https://github.com/Kong/kongponents/commit/26cfe34b6fd9fbb3d61ffdb95412804bac88ba36))





## 6.1.15 (2021-11-19)


### Bug Fixes

* **build:** update lerna ignore flag ([#431](https://github.com/Kong/kongponents/issues/431)) ([768c5e4](https://github.com/Kong/kongponents/commit/768c5e4e78bfca118fe7b2d022778854e076b323))





## 6.1.14 (2021-11-19)

**Note:** Version bump only for package @kongponents/kcard





## 6.1.13 (2021-11-19)

**Note:** Version bump only for package @kongponents/kcard





## 6.1.12 (2021-11-19)

**Note:** Version bump only for package @kongponents/kcard





## 6.1.11 (2021-11-19)

**Note:** Version bump only for package @kongponents/kcard





## 6.1.10 (2021-11-18)

**Note:** Version bump only for package @kongponents/kcard





## 6.1.9 (2021-11-18)

**Note:** Version bump only for package @kongponents/kcard





## 6.1.8 (2021-11-18)


### Bug Fixes

* **css:** box-shadow does not render on ios ([#420](https://github.com/Kong/kongponents/issues/420)) ([c0ba7cb](https://github.com/Kong/kongponents/commit/c0ba7cb90bf3c158a1baa6b879d33230359db830))





## 6.1.7 (2021-11-17)

**Note:** Version bump only for package @kongponents/kcard





## 6.1.6 (2021-11-17)


### Bug Fixes

* **ktextarea:** limit exceeded event ([#418](https://github.com/Kong/kongponents/issues/418)) ([f7cf421](https://github.com/Kong/kongponents/commit/f7cf421a4ded7f5b99eaa5a18d02a27d450020b7))





## 6.1.5 (2021-11-17)

**Note:** Version bump only for package @kongponents/kcard





## 6.1.4 (2021-11-17)


### Bug Fixes

* general fixes ([#415](https://github.com/Kong/kongponents/issues/415)) ([5b4bb48](https://github.com/Kong/kongponents/commit/5b4bb484d74e75779d39cf901b2cdaa23818e8e7))





## 6.1.3 (2021-11-17)

**Note:** Version bump only for package @kongponents/kcard





## [6.1.2](https://github.com/Kong/kongponents/compare/@kongponents/kcard@5.0.8...@kongponents/kcard@6.1.2) (2021-11-17)

### BREAKING CHANGES
  * `helpText` property removed

### Features
  * `status` / `statusHat` prop and slot added 
  * `notifications` slot added

## [5.0.8](https://github.com/Kong/kongponents/compare/v5.0.7...v5.0.8) (2021-11-15)

**Note:** Version bump only for package @kongponents/kcard





## [5.0.7](https://github.com/Kong/kongponents/compare/v5.0.6...v5.0.7) (2021-10-26)

**Note:** Version bump only for package @kongponents/kcard





## [5.0.6](https://github.com/Kong/kongponents/compare/v5.0.5...v5.0.6) (2021-10-26)

**Note:** Version bump only for package @kongponents/kcard





## [5.0.5](https://github.com/Kong/kongponents/compare/v5.0.4...v5.0.5) (2021-10-13)

**Note:** Version bump only for package @kongponents/kcard





## [5.0.4](https://github.com/Kong/kongponents/compare/v5.0.3...v5.0.4) (2021-10-06)

**Note:** Version bump only for package @kongponents/kcard





## [5.0.3](https://github.com/Kong/kongponents/compare/v5.0.2...v5.0.3) (2021-09-22)

**Note:** Version bump only for package @kongponents/kcard





## [5.0.2](https://github.com/Kong/kongponents/compare/v5.0.1...v5.0.2) (2021-09-15)

**Note:** Version bump only for package @kongponents/kcard





## [5.0.1](https://github.com/Kong/kongponents/compare/v5.0.0...v5.0.1) (2021-09-13)

**Note:** Version bump only for package @kongponents/kcard





# [5.0.0](https://github.com/Kong/kongponents/compare/v4.0.2...v5.0.0) (2021-09-13)

**Note:** Version bump only for package @kongponents/kcard





## [4.0.2](https://github.com/Kong/kongponents/compare/v4.0.1...v4.0.2) (2021-09-09)

**Note:** Version bump only for package @kongponents/kcard





## [4.0.1](https://github.com/Kong/kongponents/compare/v4.0.0...v4.0.1) (2021-09-07)

**Note:** Version bump only for package @kongponents/kcard





# 4.0.0 (2021-09-03)


### Bug Fixes

* **dist:** dist to all packages ([#152](https://github.com/Kong/kongponents/issues/152)) ([217c46b](https://github.com/Kong/kongponents/commit/217c46b02afe960f8fa597ba77410dcf034b1716))
* **kbutton:** change height of kbutton icons to be auto - intf-2282 ([#185](https://github.com/Kong/kongponents/issues/185)) ([6f497ba](https://github.com/Kong/kongponents/commit/6f497bab9a26007b51572f0403ab6ce5225f901c))
* **KCard:** add bottom margin to help text element ([#313](https://github.com/Kong/kongponents/issues/313)) ([744dd02](https://github.com/Kong/kongponents/commit/744dd02c8822251bbecdff920d25c3a541fd481a))
* **KCard:** adjust helpText margins ([#315](https://github.com/Kong/kongponents/issues/315)) ([1a09ccd](https://github.com/Kong/kongponents/commit/1a09ccde5e547bc1b78716e3b0648bb65ed388b6))
* **KCard:** Remove default cursor pointer & version ([#35](https://github.com/Kong/kongponents/issues/35)) ([cf01828](https://github.com/Kong/kongponents/commit/cf018282af74c85774db862025a71f23426e3c10))


### Features

* **kcard:** add permanent shadow property ([#188](https://github.com/Kong/kongponents/issues/188)) ([37dbee7](https://github.com/Kong/kongponents/commit/37dbee7cdac5c99597d02f54cbb25dce54e689ac))
* **KCard:** add help text prop ([#309](https://github.com/Kong/kongponents/issues/309)) ([fc13244](https://github.com/Kong/kongponents/commit/fc132445189be95da6e84ad490d28305b3f3572b))
* **style:** Updates Colors & Component Styles ([#225](https://github.com/Kong/kongponents/issues/225)) ([642a26b](https://github.com/Kong/kongponents/commit/642a26b888cbf8b4abbe32965729d44d54f9dfef))


### BREAKING CHANGES

* **style:** All new colors and color naming. Various component style improvements.





## [1.1.2](https://github.com/Kong/kongponents/compare/@kongponents/kcard@1.1.1...@kongponents/kcard@1.1.2) (2021-07-07)


### Bug Fixes

* **KCard:** adjust helpText margins ([#315](https://github.com/Kong/kongponents/issues/315)) ([1a09ccd](https://github.com/Kong/kongponents/commit/1a09ccde5e547bc1b78716e3b0648bb65ed388b6))





## [1.1.1](https://github.com/Kong/kongponents/compare/@kongponents/kcard@1.1.0...@kongponents/kcard@1.1.1) (2021-07-01)


### Bug Fixes

* **KCard:** add bottom margin to help text element ([#313](https://github.com/Kong/kongponents/issues/313)) ([744dd02](https://github.com/Kong/kongponents/commit/744dd02c8822251bbecdff920d25c3a541fd481a))





# [1.1.0](https://github.com/Kong/kongponents/compare/@kongponents/kcard@1.0.2...@kongponents/kcard@1.1.0) (2021-06-29)


### Features

* **KCard:** add help text prop ([#309](https://github.com/Kong/kongponents/issues/309)) ([fc13244](https://github.com/Kong/kongponents/commit/fc132445189be95da6e84ad490d28305b3f3572b))





## [1.0.2](https://github.com/Kong/kongponents/compare/@kongponents/kcard@1.0.1...@kongponents/kcard@1.0.2) (2021-03-10)

**Note:** Version bump only for package @kongponents/kcard





## [1.0.1](https://github.com/Kong/kongponents/compare/@kongponents/kcard@1.0.0...@kongponents/kcard@1.0.1) (2021-02-26)

**Note:** Version bump only for package @kongponents/kcard





# [1.0.0](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.2.3...@kongponents/kcard@1.0.0) (2020-06-25)


### Features

* **style:** Updates Colors & Component Styles ([#225](https://github.com/Kong/kongponents/issues/225)) ([642a26b](https://github.com/Kong/kongponents/commit/642a26b888cbf8b4abbe32965729d44d54f9dfef))


### BREAKING CHANGES

* **style:** All new colors and color naming. Various component style improvements.





## [0.2.3](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.2.2...@kongponents/kcard@0.2.3) (2020-04-24)

**Note:** Version bump only for package @kongponents/kcard





## [0.2.2](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.2.1...@kongponents/kcard@0.2.2) (2020-03-25)

**Note:** Version bump only for package @kongponents/kcard





## [0.2.1](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.2.0...@kongponents/kcard@0.2.1) (2020-03-06)

**Note:** Version bump only for package @kongponents/kcard





# [0.2.0](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.1.8...@kongponents/kcard@0.2.0) (2020-01-24)


### Features

* **kcard:** add permanent shadow property ([#188](https://github.com/Kong/kongponents/issues/188)) ([37dbee7](https://github.com/Kong/kongponents/commit/37dbee7cdac5c99597d02f54cbb25dce54e689ac))





## [0.1.8](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.1.6...@kongponents/kcard@0.1.8) (2020-01-22)

**Note:** Version bump only for package @kongponents/kcard





## [0.1.5](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.1.3-7...@kongponents/kcard@0.1.5) (2020-01-16)

**Note:** Version bump only for package @kongponents/kcard





## [0.1.3-7](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.1.3-6...@kongponents/kcard@0.1.3-7) (2019-11-12)

**Note:** Version bump only for package @kongponents/kcard





## [0.1.3-6](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.1.3-5...@kongponents/kcard@0.1.3-6) (2019-11-11)

**Note:** Version bump only for package @kongponents/kcard





## [0.1.3-5](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.1.3-4...@kongponents/kcard@0.1.3-5) (2019-11-08)


### Bug Fixes

* **dist:** dist to all packages ([#152](https://github.com/Kong/kongponents/issues/152)) ([217c46b](https://github.com/Kong/kongponents/commit/217c46b02afe960f8fa597ba77410dcf034b1716))





## [0.1.3-4](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.1.3-3...@kongponents/kcard@0.1.3-4) (2019-11-08)

**Note:** Version bump only for package @kongponents/kcard





## [0.1.3-3](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.1.3-1...@kongponents/kcard@0.1.3-3) (2019-11-06)

**Note:** Version bump only for package @kongponents/kcard





## [0.1.3-2](https://github.com/Kong/kongponents/compare/@kongponents/kcard@0.1.3-1...@kongponents/kcard@0.1.3-2) (2019-11-06)

**Note:** Version bump only for package @kongponents/kcard

### 0.1.3-1
 - check `title` slot in `$slots` for backwards compat with Vue version < 2.6 [#120](https://github.com/Kong/kongponents/pull/120)

### 0.1.3
 - add `body` only support [#108](https://github.com/Kong/kongponents/pull/108)
