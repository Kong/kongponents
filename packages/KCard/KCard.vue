<template>
  <div
    :class="[borderVariant, {'hover': hasHover }]"
    class="kong-card">
    <div class="k-card-header">
      <div class="k-card-title">
        <h4>
          <!-- @slot Use this slot to pass title content -->
          <slot name="title">{{ title }}</slot>
        </h4>
      </div>
      <div class="k-card-actions">
        <!-- @slot Use this slot to pass actions to right side of header -->
        <slot name="actions"/>
      </div>
    </div>
    <div class="k-card-body">
      <!-- @slot Use this slot to pass in body content -->
      <slot name="body">{{ body }}</slot>
    </div>
  </div>
</template>

<script>
export default {
  name: 'KCard',

  props: {
    /**
     * Pass title sting in if slot not used
     */
    title: {
      type: String,
      default: ''
    },
    /**
     * Pass body sting in if slot not used
     */
    body: {
      type: String,
      default: ''
    },
    /**
      * Set top border or no border. If neither set defaut will have border<br>
      * Options: [borderTop, noBorder]
      */
    borderVariant: {
      type: String,
      default: 'border'
    },
    /**
      * Sets if card has hover state<br>
      */
    hasHover: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style scoped>
.kong-card {
  padding: 1rem;
  margin-bottom: 1rem;
}

.kong-card.noBoard {
  border: none;
}

.kong-card.border {
  border-radius: 3px;
  border: 1px solid rgba(0,0,0,.12);
  box-shadow: none;
}

.kong-card.borderTop {
  border-top: 1px solid rgba(0, 0, 0, 0.08);
}

.kong-card.hover:hover {
  box-shadow: 0 4px 8px rgba(0,0,0,.12);
}

.kong-card .k-card-header {
  display: flex;
  align-items: center;
  margin-bottom: 1rem;
}

.kong-card .k-card-title h4 {
  margin: 0;
  font-size: 18px;
  font-weight: 500;
  color: rgba(0,0,0,.85);
}

.kong-card .k-card-actions  {
  margin-left: auto;
}
</style>
