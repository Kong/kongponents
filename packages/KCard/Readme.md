```js
<KCard>
  <template slot="title">Title</template>
  <template slot="actions"><a href="">Link</a></template>
  <template slot="body">
    <p style="margin: 0">Some kind of content can go here.</p>
  </template>
</KCard>  

```
