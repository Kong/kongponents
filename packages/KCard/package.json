{
  "name": "@kongponents/kcard",
  "version": "5.0.1",
  "description": "Card component",
  "main": "dist/KCard.umd.min.js",
  "componentName": "KCard",
  "source": "KCard.vue",
  "files": [
    "dist",
    "KCard.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  }
}
