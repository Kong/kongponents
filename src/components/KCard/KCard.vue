<template>
  <section
    :aria-describedby="contentId || undefined"
    :aria-label="title ? title : undefined"
    :aria-labelledby="!title && slots.title ? titleId : undefined"
    class="kong-card"
    :class="[borderVariant, {'hover': hasHover, 'kcard-shadow': hasShadow }]"
  >
    <div
      v-if="showCardHead"
      class="k-card-header d-flex mb-3"
      :class="{ 'has-status': status || slots.statusHat }"
    >
      <div
        v-if="status || slots.statusHat"
        class="k-card-status-hat"
      >
        <!-- @slot Use this slot to pass status text above title -->
        <slot name="statusHat">
          {{ status }}
        </slot>
      </div>

      <div
        v-if="showCardTitleWithoutStatus"
        :id="title ? undefined : titleId"
        class="k-card-title mb-3"
      >
        <h4>
          <!-- @slot Use this slot to pass title content -->
          <slot name="title">
            {{ title }}
          </slot>
        </h4>
      </div>

      <div class="k-card-actions">
        <!-- @slot Use this slot to pass actions to right side of header -->
        <slot name="actions" />
      </div>
    </div>

    <div
      v-if="showCardTitleWithStatus"
      :id="title ? undefined : titleId"
      class="k-card-title mb-3"
    >
      <h4>
        <!-- @slot Use this slot to pass title content -->
        <slot name="title">
          {{ title }}
        </slot>
      </h4>
    </div>

    <div class="k-card-content d-flex">
      <div
        :id="contentId"
        class="k-card-body"
      >
        <!-- @slot Use this slot to pass in body content -->
        <slot name="body">
          {{ body }}
        </slot>
      </div>

      <div
        v-if="slots.notifications"
        class="k-card-notifications ml-3"
      >
        <slot name="notifications" />
      </div>
    </div>
  </section>
</template>

<script lang="ts" setup>
import { v1 as uuidv1 } from 'uuid'
import { BorderVariant, BorderVariantsArray } from '@/types'
import { computed, PropType, useSlots } from 'vue'

const props = defineProps({
  /**
   * Title string if slot not used, also used for aria-label
   */
  title: {
    type: String,
    default: '',
  },
  /**
   * Pass body string in if slot not used
   */
  body: {
    type: String,
    default: '',
  },
  /**
   * Set top border or no border. If neither set default will have border<br>
   * * Options: [borderTop, noBorder]
   */
  borderVariant: {
    type: String as PropType<BorderVariant>,
    default: 'border',
    validator: (value: BorderVariant): boolean => Object.values(BorderVariantsArray).includes(value),
  },
  /**
   * Sets if card has hover state<br>
   */
  hasHover: {
    type: Boolean,
    default: false,
  },
  hasShadow: {
    type: Boolean,
    default: false,
  },
  /**
   * Add small status text above the card title
   */
  status: {
    type: String,
    default: '',
  },
  /**
   * Test mode - for testing only, strips out generated ids
   */
  testMode: {
    type: Boolean,
    default: false,
  },
})

const titleId = computed((): string => props.testMode ? 'test-title-id-1234' : uuidv1())

const contentId = computed((): string => props.testMode ? 'test-content-id-1234' : uuidv1())

const slots = useSlots()

const useStatusHatLayout = computed((): boolean => !!(props.status || !!slots.statusHat))

const showCardHead = computed((): boolean => !!slots.actions || useStatusHatLayout.value ||
  (!useStatusHatLayout.value && (!!props.title || !!slots.title)))

const showCardTitleWithoutStatus = computed((): boolean => !useStatusHatLayout.value &&
  (!!props.title || !!slots.title))

const showCardTitleWithStatus = computed((): boolean => useStatusHatLayout.value &&
  (!!props.title || !!slots.title))
</script>

<style lang="scss" scoped>
@import '@/styles/variables';
@import '@/styles/functions';

.kong-card {
  background-color: var(--KCardBackground, var(--white, color(white)));
  border-radius: var(--KCardBorderRadius, 3px);
  padding: var(--KCardPaddingY, var(--spacing-lg)) var(--KCardPaddingX, var(--spacing-lg));
  transition: box-shadow 0.2s ease-in-out;

  &.noBorder {
    border: none;
  }

  &.border {
    border: var(--KCardBorder, 1px solid var(--black-10, color(black-10)));
    box-shadow: none;
  }

  &.borderTop {
    border-top: var(--KCardBorder, 1px solid var(--black-10, color(black-10)));
  }

  &.hover:hover, &.kcard-shadow {
    box-shadow: var(--KCardShadow, 0 4px 8px var(--black-10, color(black-10)));
  }

  .k-card-header {
    align-items: center;

    &.has-status {
      align-items: flex-start;
    }
  }

  .k-card-status-hat {
    align-items: center;
    color: var(--grey-600);
    display: flex;
    font-size: var(--type-xs);
  }

  .k-card-title h4 {
    color: var(--KCardTitleColor, var(--black-500));
    font-size: var(--KCardTitleFontSize, 20px);
    font-weight: 600;
    margin: 0;
    padding: 0;
  }

  .k-card-actions  {
    margin-left: auto;
  }

  .k-card-body {
    color: var(--grey-600);
    font-size: 13px;
    line-height: 20px;
    width: 100%;
  }

  .k-card-content .k-table td,
  .k-card-content table td,
  .k-card-content :deep(.k-table) td {
    font-size: var(--type-md);
  }

  .k-card-notifications {
    margin-left: auto;
    margin-top: auto;
  }
}
</style>

<style lang="scss">
@import '@/styles/variables';
@import '@/styles/functions';

.kong-card {
  .k-card-header {
    .k-button {
      min-height: 38px;
    }
  }
}
</style>
