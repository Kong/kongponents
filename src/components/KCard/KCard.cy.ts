import { mount } from 'cypress/vue'
import KCard from '@/components/KCard/KCard.vue'

describe('KCard', () => {
  it('renders empty card element when no props or slots are passed', () => {
    mount(KCard)

    cy.get('.k-card').should('be.visible')
    cy.get('.k-card').find('card-header').should('not.exist')
    cy.get('.k-card').find('card-content').should('not.exist')
    cy.get('.k-card').find('card-footer').should('not.exist')
  })

  it('renders title prop when passed', () => {
    const titleProp = 'Title prop'

    mount(KCard, {
      props: {
        title: titleProp,
      },
    })

    cy.get('.k-card').find('.card-title').should('contain', titleProp)
  })

  it('renders slots when passed', () => {
    const titleProp = 'Test title'
    const titleText = 'I am the title'

    mount(KCard, {
      props: {
        title: titleProp,
      },
      slots: {
        title: `<span data-testid="card-title">${titleText}</span>`,
        actions: '<span data-testid="card-actions">Card actions</span>',
        default: '<span data-testid="card-content">Card content</span>',
        footer: '<span data-testid="card-footer">Card footer</span>',
      },
    })

    cy.get('.k-card').find('.card-title').get('[data-testid="card-title"]').should('contain', titleText)
    cy.get('.k-card').find('.card-actions').get('[data-testid="card-actions"]').should('be.visible')
    cy.get('.k-card').find('.card-content').get('[data-testid="card-content"]').should('be.visible')
    cy.get('.k-card').find('.card-footer').get('[data-testid="card-footer"]').should('be.visible')
  })
})
