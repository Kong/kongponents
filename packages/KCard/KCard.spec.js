import { mount } from '@vue/test-utils'
import KCard from '@/KCard/KCard'

describe('KCard', () => {
  it('renders slots when passed', () => {
    const cardTitle = 'Card Title'
    const cardBody = 'Card Body'
    const wrapper = mount(KCard, {
      slots: {
        'title': `<span>${cardTitle}</span>`,
        'body': `<div>${cardBody}</div>`
      }
    })

    expect(wrapper.find('.k-card-title').html()).toEqual(expect.stringContaining(cardTitle))
    expect(wrapper.find('.k-card-body').html()).toEqual(expect.stringContaining(cardBody))
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('has border class when passed', () => {
    const wrapper = mount(KCard, {
      propsData: {
        hasBorder: true
      }
    })

    expect(wrapper.classes()).toContain('border')
  })

  it('has hover class when passed', () => {
    const wrapper = mount(KCard, {
      propsData: {
        hasHover: true
      }
    })

    expect(wrapper.classes()).toContain('hover')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('matches snapshot', () => {
    const wrapper = mount(KCard)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
