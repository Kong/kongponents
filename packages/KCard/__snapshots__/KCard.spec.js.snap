// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KCard has hover class when passed 1`] = `
<div class="kong-card border hover">
  <div class="k-card-header">
    <div class="k-card-title">
      <h4></h4>
    </div>
    <div class="k-card-actions"></div>
  </div>
  <div class="k-card-body"></div>
</div>
`;

exports[`KCard matches snapshot 1`] = `
<div class="kong-card border">
  <div class="k-card-header">
    <div class="k-card-title">
      <h4></h4>
    </div>
    <div class="k-card-actions"></div>
  </div>
  <div class="k-card-body"></div>
</div>
`;

exports[`KCard renders slots when passed 1`] = `
<div class="kong-card border">
  <div class="k-card-header">
    <div class="k-card-title">
      <h4><span>Card Title</span></h4>
    </div>
    <div class="k-card-actions"></div>
  </div>
  <div class="k-card-body">
    <div>Card Body</div>
  </div>
</div>
`;
